File: golang/e2e/cyrillic_e2e_test.go
package e2e_test

import (
	"context"
	"database/sql"
	"encoding/json"
	"fmt"
	"os"
	"testing"

	clickhousedriver "github.com/ClickHouse/clickhouse-go/v2"
	flyql "github.com/iamtelescope/flyql/golang"
	clickhousegen "github.com/iamtelescope/flyql/golang/generators/clickhouse"
	postgresqlgen "github.com/iamtelescope/flyql/golang/generators/postgresql"
	starrocksgen "github.com/iamtelescope/flyql/golang/generators/starrocks"
	"github.com/iamtelescope/flyql/golang/matcher"
	"github.com/jackc/pgx/v5"
)

// cyrillicTable is a dedicated table seeded with Cyrillic data (see each
// tests-data/e2e/<db>/init.sql). Kept separate from flyql_e2e_test so the
// shared 6-row dataset and its expected_ids stay untouched.
const cyrillicTable = "flyql_cyrillic_test"

type cyrillicFixture struct {
	Rows  []map[string]any `json:"rows"`
	Tests []testCase       `json:"tests"`
}

func loadCyrillicFixture(t *testing.T) cyrillicFixture {
	t.Helper()
	data, err := os.ReadFile(testDataPath("cyrillic.json"))
	if err != nil {
		t.Fatalf("read cyrillic.json: %v", err)
	}
	var f cyrillicFixture
	if err := json.Unmarshal(data, &f); err != nil {
		t.Fatalf("parse cyrillic.json: %v", err)
	}
	return f
}

// TestCyrillicMatcherE2E evaluates every fixture case with the in-memory matcher.
// Identical expected_ids across Python/Go/JS prove cross-language parsing parity
// for non-ASCII input — no database required.
func TestCyrillicMatcherE2E(t *testing.T) {
	fx := loadCyrillicFixture(t)
	for _, tc := range fx.Tests {
		tc := tc
		t.Run(tc.Name, func(t *testing.T) {
			r := testResult{
				Kind:        "where",
				Database:    "matcher",
				Name:        tc.Name,
				FlyQL:       tc.FlyQL,
				SQL:         "(in-memory)",
				ExpectedIDs: tc.ExpectedIDs,
			}

			var matchedIDs []int
			for _, row := range fx.Rows {
				matched, err := matcher.Match(tc.FlyQL, row)
				if err != nil {
					r.Error = err.Error()
					addResult(r)
					t.Fatalf("matcher error: %v", err)
				}
				if matched {
					// JSON unmarshals numbers as float64; fail loudly on anything else
					// rather than silently dropping the row from the result set.
					id, ok := row["id"].(float64)
					if !ok {
						r.Error = fmt.Sprintf("row id is not a JSON number: %T", row["id"])
						addResult(r)
						t.Fatalf("row id is not a JSON number: %T", row["id"])
					}
					matchedIDs = append(matchedIDs, int(id))
				}
			}

			r.ReturnedIDs = matchedIDs
			r.Passed = idsMatch(tc.ExpectedIDs, matchedIDs)
			addResult(r)
			if !r.Passed {
				t.Errorf("expected %v, got %v", tc.ExpectedIDs, matchedIDs)
			}
		})
	}
}

// TestCyrillicDatabaseE2E generates SQL from each fixture case and runs it against
// the Cyrillic table in every database the case lists. Each database is skipped
// independently when unavailable. The e2e runner cross-checks that Python/Go/JS
// produced byte-identical SQL per (case, database).
func TestCyrillicDatabaseE2E(t *testing.T) {
	fx := loadCyrillicFixture(t)
	ctx := context.Background()

	// ClickHouse — open, always defer Close (even if Ping fails), then probe.
	chConn, chErr := clickhousedriver.Open(&clickhousedriver.Options{
		Addr: []string{clickhouseAddr()},
		Auth: clickhousedriver.Auth{Database: "default", Username: "flyql", Password: "flyql"},
	})
	chAvail := false
	if chErr == nil {
		defer chConn.Close()
		chAvail = chConn.Ping(ctx) == nil
	}

	// StarRocks
	srDB, srErr := sql.Open("mysql", starRocksDSN())
	srAvail := false
	if srErr == nil {
		defer srDB.Close()
		srAvail = srDB.Ping() == nil
	}

	// PostgreSQL — pgx.Connect alone doesn't prove the conn is usable; Ping it.
	pgConn, pgErr := pgx.Connect(ctx, postgresqlDSN())
	pgAvail := false
	if pgErr == nil {
		defer pgConn.Close(ctx)
		pgAvail = pgConn.Ping(ctx) == nil
	}

	if !chAvail && !srAvail && !pgAvail {
		t.Skip("no database available for Cyrillic e2e")
	}

	chCols := loadClickHouseColumns(t)
	srCols := loadStarRocksColumns(t)
	pgCols := loadPostgreSQLColumns(t)

	runCH := func(expr string) (string, []int, error) {
		parsed, err := flyql.Parse(expr)
		if err != nil {
			return "", nil, fmt.Errorf("parse: %w", err)
		}
		sqlWhere, err := clickhousegen.ToSQLWhere(parsed.Root, chCols)
		if err != nil {
			return "", nil, fmt.Errorf("generate: %w", err)
		}
		rows, err := chConn.Query(ctx, fmt.Sprintf("SELECT id FROM %s WHERE %s ORDER BY id", cyrillicTable, sqlWhere))
		if err != nil {
			return sqlWhere, nil, fmt.Errorf("query: %w", err)
		}
		defer rows.Close()
		var ids []int
		for rows.Next() {
			var id int32
			if err := rows.Scan(&id); err != nil {
				return sqlWhere, nil, fmt.Errorf("scan: %w", err)
			}
			ids = append(ids, int(id))
		}
		return sqlWhere, ids, rows.Err()
	}

	runSR := func(expr string) (string, []int, error) {
		parsed, err := flyql.Parse(expr)
		if err != nil {
			return "", nil, fmt.Errorf("parse: %w", err)
		}
		sqlWhere, err := starrocksgen.ToSQLWhere(parsed.Root, srCols)
		if err != nil {
			return "", nil, fmt.Errorf("generate: %w", err)
		}
		rows, err := srDB.Query(fmt.Sprintf("SELECT id FROM %s WHERE %s ORDER BY id", cyrillicTable, sqlWhere))
		if err != nil {
			return sqlWhere, nil, fmt.Errorf("query: %w", err)
		}
		defer rows.Close()
		var ids []int
		for rows.Next() {
			var id int32
			if err := rows.Scan(&id); err != nil {
				return sqlWhere, nil, fmt.Errorf("scan: %w", err)
			}
			ids = append(ids, int(id))
		}
		return sqlWhere, ids, rows.Err()
	}

	runPG := func(expr string) (string, []int, error) {
		parsed, err := flyql.Parse(expr)
		if err != nil {
			return "", nil, fmt.Errorf("parse: %w", err)
		}
		sqlWhere, err := postgresqlgen.ToSQLWhere(parsed.Root, pgCols)
		if err != nil {
			return "", nil, fmt.Errorf("generate: %w", err)
		}
		rows, err := pgConn.Query(ctx, fmt.Sprintf("SELECT id FROM %s WHERE %s ORDER BY id", cyrillicTable, sqlWhere))
		if err != nil {
			return sqlWhere, nil, fmt.Errorf("query: %w", err)
		}
		defer rows.Close()
		var ids []int
		for rows.Next() {
			var id int32
			if err := rows.Scan(&id); err != nil {
				return sqlWhere, nil, fmt.Errorf("scan: %w", err)
			}
			ids = append(ids, int(id))
		}
		return sqlWhere, ids, rows.Err()
	}

	dbs := []struct {
		name  string
		avail bool
		run   func(string) (string, []int, error)
	}{
		{"clickhouse", chAvail, runCH},
		{"starrocks", srAvail, runSR},
		{"postgresql", pgAvail, runPG},
	}

	for _, tc := range fx.Tests {
		tc := tc
		t.Run(tc.Name, func(t *testing.T) {
			for _, db := range dbs {
				if !containsDB(tc.Databases, db.name) || !db.avail {
					continue
				}
				r := testResult{
					Kind:        "where",
					Database:    db.name,
					Name:        tc.Name,
					FlyQL:       tc.FlyQL,
					ExpectedIDs: tc.ExpectedIDs,
				}
				sqlText, ids, runErr := db.run(tc.FlyQL)
				r.SQL = sqlText
				if runErr != nil {
					r.Error = runErr.Error()
					addResult(r)
					t.Errorf("%s: %v", db.name, runErr)
					continue
				}
				r.ReturnedIDs = ids
				r.Passed = idsMatch(tc.ExpectedIDs, ids)
				addResult(r)
				if !r.Passed {
					t.Errorf("%s: expected %v got %v", db.name, tc.ExpectedIDs, ids)
				}
			}
		})
	}
}

File: javascript/e2e/cyrillic.e2e.test.js
import { describe, it, beforeAll, afterAll, expect } from 'vitest'
import fs from 'fs'
import path from 'path'
import { fileURLToPath } from 'url'
import { match } from 'flyql/matcher'
import { parse } from 'flyql/core'
import { generateWhere as chGenerateWhere, newColumn as chNewColumn } from 'flyql/generators/clickhouse'
import { generateWhere as srGenerateWhere, newColumn as srNewColumn } from 'flyql/generators/starrocks'
import { generateWhere as pgGenerateWhere, newColumn as pgNewColumn } from 'flyql/generators/postgresql'

// Cross-language Cyrillic (non-ASCII) e2e. Driven by tests-data/e2e/cyrillic.json:
// every case is evaluated in-memory (matcher) and, when listed, against the
// dedicated flyql_cyrillic_test table in each database. Identical expected_ids
// across Python/Go/JS prove cross-language parity. Regression coverage for the
// byte-vs-codepoint offset bug in the Go parser.

const __filename = fileURLToPath(import.meta.url)
const __dirname = path.dirname(__filename)
const testDataDir = path.join(__dirname, '..', '..', 'tests-data', 'e2e')

const CYRILLIC_TABLE = 'flyql_cyrillic_test'

const CH_HOST = process.env.CLICKHOUSE_HOST || 'localhost'
const CH_PORT = process.env.CLICKHOUSE_HTTP_PORT || '18123'
const CH_USER = process.env.CLICKHOUSE_USER || 'flyql'
const CH_PASS = process.env.CLICKHOUSE_PASSWORD || 'flyql'

const SR_HOST = process.env.STARROCKS_HOST || 'localhost'
const SR_HTTP_PORT = process.env.STARROCKS_HTTP_PORT || '18030'
const SR_USER = process.env.STARROCKS_USER || 'root'
const SR_PASS = process.env.STARROCKS_PASSWORD || ''

const PG_HOST = process.env.POSTGRESQL_HOST || 'localhost'
const PG_PORT = process.env.POSTGRESQL_PORT || '15432'
const PG_USER = process.env.POSTGRESQL_USER || 'flyql'
const PG_PASS = process.env.POSTGRESQL_PASSWORD || 'flyql'
const PG_DB = process.env.POSTGRESQL_DB || 'flyql_test'

const REPORT_PATH = process.env.E2E_REPORT_JSON || ''
const reportResults = []

function loadJSON(filepath) {
    return JSON.parse(fs.readFileSync(filepath, 'utf-8'))
}

const fixture = loadJSON(path.join(testDataDir, 'cyrillic.json'))
const rows = fixture.rows
const cases = fixture.tests

function buildColumns(dialect, factory) {
    const colData = loadJSON(path.join(testDataDir, dialect, 'columns.json'))
    const columns = {}
    for (const [key, col] of Object.entries(colData.columns)) {
        columns[key] = factory({ name: col.name, type: col.type, values: col.values })
    }
    return columns
}

async function chQuery(sql) {
    const params = new URLSearchParams({ user: CH_USER, password: CH_PASS, default_format: 'JSONEachRow' })
    const response = await fetch(`http://${CH_HOST}:${CH_PORT}/?${params}`, { method: 'POST', body: sql })
    const text = await response.text()
    if (!response.ok) throw new Error(`ClickHouse error: ${text.trim()}`)
    if (!text.trim()) return []
    return text.trim().split('\n').map((line) => JSON.parse(line))
}

async function srQuery(sql) {
    const credentials = Buffer.from(`${SR_USER}:${SR_PASS}`).toString('base64')
    const url = `http://${SR_HOST}:${SR_HTTP_PORT}/api/v1/catalogs/default_catalog/databases/flyql_test/sql`
    const response = await fetch(url, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json', Authorization: `Basic ${credentials}` },
        body: JSON.stringify({ query: `${sql};` }),
    })
    const text = await response.text()
    if (!response.ok) throw new Error(`StarRocks HTTP error ${response.status}: ${text}`)
    const lines = text.trim().split('\n').filter((l) => l.trim()).map((l) => JSON.parse(l))
    let colNames = []
    const out = []
    for (const obj of lines) {
        if (obj.meta) colNames = obj.meta.map((m, i) => m.name || `col${i}`)
        else if (obj.data) {
            const row = {}
            colNames.forEach((n, i) => {
                row[n] = obj.data[i]
            })
            out.push(row)
        }
    }
    return out
}

async function pgQuery(sql) {
    const { execSync } = await import('child_process')
    const env = {
        ...process.env,
        PGHOST: PG_HOST,
        PGPORT: PG_PORT,
        PGUSER: PG_USER,
        PGPASSWORD: PG_PASS,
        PGDATABASE: PG_DB,
    }
    const result = execSync(`psql -t -A -F '\t' -c ${JSON.stringify(sql)}`, {
        env,
        encoding: 'utf-8',
        timeout: 10000,
    })
    return result.trim().split('\n').filter((l) => l.trim() !== '')
}

async function runClickHouse(flyqlExpr, columns) {
    const sqlWhere = chGenerateWhere(parse(flyqlExpr).root, columns)
    const result = await chQuery(`SELECT id FROM ${CYRILLIC_TABLE} WHERE ${sqlWhere} ORDER BY id`)
    return { sql: sqlWhere, ids: result.map((r) => Number(r.id)) }
}

async function runStarRocks(flyqlExpr, columns) {
    const sqlWhere = srGenerateWhere(parse(flyqlExpr).root, columns)
    const result = await srQuery(`SELECT id FROM ${CYRILLIC_TABLE} WHERE ${sqlWhere} ORDER BY id`)
    return { sql: sqlWhere, ids: result.map((r) => Number(r.id)) }
}

async function runPostgreSQL(flyqlExpr, columns) {
    const sqlWhere = pgGenerateWhere(parse(flyqlExpr).root, columns)
    const lines = await pgQuery(`SELECT id FROM ${CYRILLIC_TABLE} WHERE ${sqlWhere} ORDER BY id`)
    return { sql: sqlWhere, ids: lines.map((line) => Number(line.trim())) }
}

function writeReport() {
    if (!REPORT_PATH || reportResults.length === 0) return
    try {
        let existing = { language: 'javascript', results: [] }
        if (fs.existsSync(REPORT_PATH)) {
            try {
                existing = JSON.parse(fs.readFileSync(REPORT_PATH, 'utf-8'))
            } catch {
                /* ignore */
            }
        }
        existing.results = [...(existing.results || []), ...reportResults]
        fs.writeFileSync(REPORT_PATH, JSON.stringify(existing, null, 2))
    } catch (e) {
        console.error(`warn: could not write e2e report: ${e.message}`)
    }
}

describe('Cyrillic Matcher E2E', () => {
    afterAll(writeReport)

    it.each(cases.map((tc) => [tc.name, tc.flyql, tc.expected_ids]))(
        '%s: %s',
        (name, flyql, expectedIds) => {
            const result = {
                kind: 'where',
                database: 'matcher',
                name,
                flyql,
                sql: '(in-memory)',
                expected_ids: expectedIds,
                returned_ids: [],
                passed: false,
                error: '',
            }
            try {
                const matchedIds = rows.filter((row) => match(flyql, row)).map((row) => row.id)
                result.returned_ids = matchedIds
                const sortedMatched = [...matchedIds].sort()
                const sortedExpected = [...expectedIds].sort()
                result.passed = JSON.stringify(sortedMatched) === JSON.stringify(sortedExpected)
                reportResults.push(result)
                expect(sortedMatched).toEqual(sortedExpected)
            } catch (e) {
                result.error = e.message
                reportResults.push(result)
                throw e
            }
        },
    )
})

describe('Cyrillic Database E2E', () => {
    let chCols
    let srCols
    let pgCols
    let chAvailable = false
    let srAvailable = false
    let pgAvailable = false

    beforeAll(async () => {
        chCols = buildColumns('clickhouse', chNewColumn)
        srCols = buildColumns('starrocks', srNewColumn)
        pgCols = buildColumns('postgresql', pgNewColumn)
        try {
            const r = await chQuery('SELECT 1 AS ok')
            chAvailable = r.length > 0 && Number(r[0].ok) === 1
        } catch {
            chAvailable = false
        }
        try {
            const r = await srQuery('SELECT 1 AS ok')
            srAvailable = r.length > 0 && Number(r[0].ok) === 1
        } catch {
            srAvailable = false
        }
        try {
            const lines = await pgQuery('SELECT 1 AS ok')
            pgAvailable = lines.length > 0 && lines[0].trim() === '1'
        } catch {
            pgAvailable = false
        }
    })

    afterAll(writeReport)

    const dbCases = cases.filter((tc) => (tc.databases || []).length > 0)

    it.each(dbCases.map((tc) => [tc.name, tc]))('%s', async (_name, tc) => {
        const expected = [...tc.expected_ids].sort((a, b) => a - b)

        const byDb = {
            clickhouse: { available: () => chAvailable, run: () => runClickHouse(tc.flyql, chCols) },
            starrocks: { available: () => srAvailable, run: () => runStarRocks(tc.flyql, srCols) },
            postgresql: { available: () => pgAvailable, run: () => runPostgreSQL(tc.flyql, pgCols) },
        }

        for (const db of tc.databases) {
            const runner = byDb[db]
            const result = {
                kind: 'where',
                database: db,
                name: tc.name,
                flyql: tc.flyql,
                sql: '',
                expected_ids: expected,
                returned_ids: [],
                passed: false,
                error: '',
            }
            if (!runner.available()) {
                // Skip silently (emit no row) so an unavailable DB matches the
                // Go (continue) and Python (pytest.skip) behavior. Pushing a
                // failing row here would make the runner's cross-language
                // collapse see a lone JS failure for a (case, db) group that
                // Go/Python omit entirely — a spurious parity mismatch.
                continue
            }
            try {
                const { sql, ids } = await runner.run()
                const sortedIds = [...ids].sort((a, b) => a - b)
                result.sql = sql
                result.returned_ids = sortedIds
                result.passed = JSON.stringify(sortedIds) === JSON.stringify(expected)
            } catch (e) {
                result.error = e.message
            }
            reportResults.push(result)
            expect(result.error, `${db}: ${result.error}`).toBe('')
            expect(result.returned_ids, `${db}`).toEqual(expected)
        }
    })
})
